sapicore/engine/network/__init__.py

```diff
@@ -0,0 +1,400 @@
+                [synapse(ensemble_ref.voltage) for synapse in outgoing_synapses]
+
+        self.simulation_step += 1
+
+        # build and pass a state dictionary for use by calling module.
+        state = {"ensembles": {}, "synapses": {}}
+        for comp in self.get_ensembles():
+            state["ensembles"][comp.identifier] = comp.loggable_state()
+
+        for comp in self.get_synapses():
+            state["synapses"][comp.identifier] = comp.loggable_state()
+
+        return state
+
+    def _in_edges(self, node: str) -> list[Synapse]:
+        """Returns list of synapse objects going into the ensemble `node`."""
+        return [ref.get("reference") for _, _, ref in list(self.graph.in_edges(node, data=True))]
+
+    def _out_edges(self, node: str) -> list[Synapse]:
+        """Returns list of synapse objects fanning out of the ensemble `node`."""
+        return [ref.get("reference") for _, _, ref in list(self.graph.out_edges(node, data=True))]
+
+    def _find_roots(self):
+        """Identifies nodes that are exposed to external input. Those are used as starting points for the multi-BFS
+        forward sweep, whose order also gets updated every time this method is called.
+
+        Warning
+        -------
+        Roots are updated on network initialization and on every application of :meth:`add_ensemble` and
+        :meth:`add_synapse`. If adding a synapse results in a rootless recurrent network, all nodes will have
+        in-degree > 0. In such cases, the method will mark the first ensemble added `self.get_ensembles()[0]`
+        as the root and direct external input to it. This behavior can be modified by setting `self.roots`
+        directly after building the network.
+
+        """
+        updated_roots = [node for node, degree in self.graph.in_degree() if degree == 0]
+        if updated_roots:
+            self.roots = updated_roots
+
+        elif self.get_ensembles():
+            # if the network became rootless after adding a synapse, default to the first ensemble as the root.
+            self.roots = [self.get_ensembles()[0].identifier]
+
+        # breadth-first search (BFS) iterator used to traverse the network graph layer by layer.
+        # returns a list of layers, first one consisting of the root nodes. Flatten and visit sequentially.
+        self.traversal_order = flatten(list(nx.bfs_layers(self.graph, self.roots)))
+
+    def _object_from_configuration(
+        self, path: str = None, cfg: dict = None, comp_type: str = None
+    ) -> (Neuron | Synapse, dict):
+        """Constructs an objects from the YAML in `path` or from `cfg` dictionary if provided."""
+        # load configuration dictionary from ensemble descriptor YAML.
+        if path and not cfg:
+            cfg = load_yaml(path)
+
+        # initialize the ensemble/synapse based on its class reference, given in its configuration file.
+        temp_import = __import__(cfg["package"], globals(), locals(), ["*"])
```

Comment:
I don't think our globals and locals need to be provided to the package being imported. So it's probably better to not pollute the package with our globals/locals!?

Also, it's a little scary to support importing any package by name as this is a huge potential security issue. But, maybe it's not a problem for us. Nonetheless, this is something that needs to be made very clear to the user, and potentially provide an allow/denylist approach.

Reply:
Disallowing dynamic imports would break the YAML network building API. Unless you want to create a whitelist by registering them to strings... but then, what about the flexibility of naming external / custom user neuron classes in a YAML?

Reply:
Exactly

Reply:
>Unless you want to create a whitelist by registering them to strings

Exactly.

>but then, what about the flexibility of naming external / custom user neuron classes in a YAML?

Right, so a user could whitelist their root packages and only modules starting with any of these root packages can be imported. But, this is something we can think about later. It may not be an issue for the forseeable future as long as you don't get config yaml files from random places.

Reply:
Resolving for now, but worth revisiting when we have less on our plate.